### PR TITLE
Modified the implementation of video_recorder to mantain frame rate output

### PR DESCRIPTION
**Pull request by zenified**
modified video_recorder to maintain framerate when video stream experience frame drops

### CODE REVIEW

File: image_view/CMakeLists.txt [hunk 1]
```diff
@@ -1,6 +1,8 @@
 cmake_minimum_required(VERSION 2.8)
 project(image_view)
 
+set(CMAKE_CXX_FLAGS "-std=c++0x ${CMAKE_CXX_FLAGS}")
```
Comment (vrabaud) [inline image_view/CMakeLists.txt]: we can't ask for a package to depend on C++11 during a release cycle. Please remove this line and have your code use proper C++03 code. Thx

File: image_view/src/nodes/video_recorder.cpp [hunk 2]
```diff
@@ -66,33 +75,76 @@ void callback(const sensor_msgs::ImageConstPtr& image_msg)
...
     }
 
-    if ((image_msg->header.stamp - g_last_wrote_time) < ros::Duration(1 / fps))
-    {
-      // Skip to get video with correct fps
-      return;
-    }
-
+    
+    // ROS_INFO_STREAM("last rec:  " << (image_msg->header.stamp - g_last_wrote_time) <<" min time: "<<ros::Duration(1.0 / fps));
```
Comment (vrabaud) [inline image_view/src/nodes/video_recorder.cpp]: remove this comment.

File: image_view/src/nodes/video_recorder.cpp [hunk 3]
```diff
@@ -66,33 +75,76 @@ void callback(const sensor_msgs::ImageConstPtr& image_msg)
...
-      const cv::Mat image = cv_bridge::cvtColorForDisplay(cv_bridge::toCvShare(image_msg), encoding, options)->image;
-      if (!image.empty()) {
-        outputVideo << image;
-        ROS_INFO_STREAM("Recording frame " << g_count << "\x1b[1F");
-        g_count++;
-        g_last_wrote_time = image_msg->header.stamp;
-      } else {
-          ROS_WARN("Frame skipped, no data!");
+      cv::Mat imageCapture = cv_bridge::cvtColorForDisplay(cv_bridge::toCvShare(image_msg), encoding, options)->image;
+
```
Comment (vrabaud) [inline image_view/src/nodes/video_recorder.cpp]: too many skipped lines, please remove

Comment (vrabaud) [reply image_view/src/nodes/video_recorder.cpp]: why has const disappeared ?

File: image_view/src/nodes/video_recorder.cpp [hunk 4]
```diff
@@ -66,33 +75,76 @@ void callback(const sensor_msgs::ImageConstPtr& image_msg)
...
-        outputVideo << image;
-        ROS_INFO_STREAM("Recording frame " << g_count << "\x1b[1F");
-        g_count++;
-        g_last_wrote_time = image_msg->header.stamp;
-      } else {
-          ROS_WARN("Frame skipped, no data!");
+      cv::Mat imageCapture = cv_bridge::cvtColorForDisplay(cv_bridge::toCvShare(image_msg), encoding, options)->image;
+      
+
+      captured++;
```
Comment (vrabaud) [inline image_view/src/nodes/video_recorder.cpp]: that variable is useless.

File: image_view/src/nodes/video_recorder.cpp [hunk 5]
```diff
@@ -66,33 +75,76 @@ void callback(const sensor_msgs::ImageConstPtr& image_msg)
...
+        {
+          outputVideo << imageCapture;
+          g_last_wrote_time = ros::Time::now();
+          g_count++;
+
+           
+          ROS_INFO_STREAM("Recording frame " << g_count << "\x1b[A");  
+          g_start_time = ros::Time::now(); 
+
+        }
```
Comment (vrabaud) [inline image_view/src/nodes/video_recorder.cpp]: } else {

File: image_view/src/nodes/video_recorder.cpp [hunk 6]
```diff
@@ -40,6 +44,11 @@ double max_depth_range;
 bool use_dynamic_range;
 int colormap;
 
+cv::Mat imageBuff;
+int called = 0;
+int captured =0;
+
+ros::NodeHandle* _nh;
```
Comment (vrabaud) [inline image_view/src/nodes/video_recorder.cpp]: what is that variable for ?

File: image_view/src/nodes/video_recorder.cpp [hunk 7]
```diff
@@ -66,33 +75,76 @@ void callback(const sensor_msgs::ImageConstPtr& image_msg)
...
+      
+
+      captured++;        
+        
+      if(!imageCapture.empty())
+      {
+        if(g_count==0)
+        {
+          outputVideo << imageCapture;
+          g_last_wrote_time = ros::Time::now();
```
Comment (vrabaud) [inline image_view/src/nodes/video_recorder.cpp]: g_last_wrote_time is now useless as it is write-only no ?

File: image_view/src/nodes/video_recorder.cpp [hunk 8]
```diff
@@ -66,33 +75,76 @@ void callback(const sensor_msgs::ImageConstPtr& image_msg)
...
+          g_start_time = ros::Time::now(); 
+
+        }
+        else
+        {
+          float elapsedTime = (ros::Time::now() - g_start_time).toSec();
+          int missingFrame = (elapsedTime*fps) - g_count;
+
+          if(missingFrame<=0){
+            imageBuff = cv_bridge::cvtColorForDisplay(cv_bridge::toCvShare(image_msg), encoding, options)->image.clone();
```
Comment (vrabaud) [inline image_view/src/nodes/video_recorder.cpp:105]: why not imageCapture.clone() ?

File: image_view/src/nodes/video_recorder.cpp [hunk 9]
```diff
@@ -40,6 +44,11 @@ double max_depth_range;
 bool use_dynamic_range;
 int colormap;
 
+cv::Mat imageBuff;
+int called = 0;
```
Comment (vrabaud) [inline image_view/src/nodes/video_recorder.cpp]: useless variable

**Review by vrabaud — CHANGES_REQUESTED** [review]
this change is way to complex for what it's supposed to do, please simplify (mostyl, reduce the number of skipped line and remove all the code you probably used for debugging).
The overall goal is good though so tthx, it's doing it the proper way !

**Comment by zenified** [comment]
Thanks for the feedback, apology for not cleaning up the code properly before the merge request.

File: image_view/src/nodes/video_recorder.cpp [hunk 10]
```diff
@@ -131,11 +173,22 @@ int main(int argc, char** argv)
         exit(-1);
     }
 
+    _nh = &nh;
+
+    int numRun = 10;
```
Comment (vrabaud) [inline image_view/src/nodes/video_recorder.cpp]: useless variable

File: image_view/src/nodes/video_recorder.cpp [hunk 11]
```diff
@@ -131,11 +173,22 @@ int main(int argc, char** argv)
...
+
+    float duration =60;
```
Comment (vrabaud) [inline image_view/src/nodes/video_recorder.cpp]: useless variable

File: image_view/src/nodes/video_recorder.cpp [hunk 12]
```diff
@@ -131,11 +173,22 @@ int main(int argc, char** argv)
...
+
+    float totalTime = 0;
```
Comment (vrabaud) [inline image_view/src/nodes/video_recorder.cpp]: useless variable

File: image_view/src/nodes/video_recorder.cpp [hunk 13]
```diff
@@ -66,33 +73,68 @@ void callback(const sensor_msgs::ImageConstPtr& image_msg)
...
+        }
+        
+      }
+
+      else {
+        ROS_WARN("Frame skipped, no data!");
       }
-    } catch(cv_bridge::Exception)
+    }
+
```
Comment (vrabaud) [inline image_view/src/nodes/video_recorder.cpp]: don't skip line

File: image_view/src/nodes/video_recorder.cpp [hunk 14]
```diff
@@ -66,33 +73,68 @@ void callback(const sensor_msgs::ImageConstPtr& image_msg)
...
+        else{
+          float elapsedTime = (ros::Time::now() - g_start_time).toSec();
+          int missingFrame = (elapsedTime*fps) - g_count;
+
+          if(missingFrame<=0){
+            imageBuff = cv_bridge::cvtColorForDisplay(cv_bridge::toCvShare(image_msg), encoding, options)->image.clone();
+              return; //skip frame
+          }
+              
+
```
Comment (vrabaud) [inline image_view/src/nodes/video_recorder.cpp]: too many

skipped

lines

below

File: image_view/src/nodes/video_recorder.cpp [hunk 15]
```diff
@@ -66,33 +73,68 @@ void callback(const sensor_msgs::ImageConstPtr& image_msg)
...
+          int imagePad = missingFrame-1;
+
+          for( int x = 0; x<imagePad;x++)
```
Comment (vrabaud) [inline image_view/src/nodes/video_recorder.cpp]: use a formatter to give something proper with spaces like: for(int x=0; x<imagePad; ++x)

File: image_view/src/nodes/video_recorder.cpp (hunk 5 above)
Comment (vrabaud) [reply image_view/src/nodes/video_recorder.cpp]: please address

File: image_view/src/nodes/video_recorder.cpp [hunk 16]
```diff
@@ -131,11 +173,22 @@ int main(int argc, char** argv)
...
     image_transport::ImageTransport it(nh);
     std::string topic = nh.resolveName("image");
     image_transport::Subscriber sub_image = it.subscribe(topic, 1, callback);
 
     ROS_INFO_STREAM("Waiting for topic " << topic << "...");
+    
     ros::spin();
-    std::cout << "\nVideo saved as " << filename << std::endl;
+
+    std::cout << "\nVideo saved to " <<filename<<  std::endl;
```
Comment (vrabaud) [inline image_view/src/nodes/video_recorder.cpp:182]: please respect spaces.

File: image_view/src/nodes/video_recorder.cpp (hunk 8 above)
Comment (vrabaud) [reply image_view/src/nodes/video_recorder.cpp:105]: please address

**Review by vrabaud — CHANGES_REQUESTED** [review]
Please use a C++ formatter and maybe a GUI to show which variables are useless.

**Comment by zenified** [comment]
I have formatted the code with clang-format as recommended in the ros style guide. The formatting issues should have all been corrected.

I have ran the code the code through a static analyzer and removed all the redundant variables  identified.

Thanks again and apologies for trouble

**Comment by vrabaud** [comment]
aaaaaaah, you should never run a formatter on the whole file, only on the section you modified :) Otherwise it makes things very hard for me to review (look at the diff size, it's crazy now). Any way you can revert your previous commit? (in Eclipse, you can only format a section, not sure about other editors).

**Comment by zenified** [comment]
ah my bad.. let me try to revert ><

**Comment by zenified** [comment]
If this revert makes the changes too messy, perhaps i could close this pull request and try to rebase properly on my end before I raise another request.

**Comment by JWhitleyWork** [comment]
@zenified - If this is still needed, please re-apply these changes in a new, cleaner PR against Melodic.

**Pull request closed without merging**